Added support for the go to definition functionality

JordyMackay opened the pull request:
- [x] onDidChangeContent is called for every single change, we don't want to find references every time the document changes (but batch them somehow)
- [x] change the name of omtLinkProvider (because it does provide more than only links), or split functionality
- [x] unittests
- [x] move yaml dependency to server
- [x] update CHANGELOG.md

RogierEnders commented:
Nice.

JordyMackay commented on server/src/server.ts:
```diff
@@ -73,6 +81,69 @@ connection.onInitialized(() => {
+    const otherDocument = TextDocument.create(linkUrl, 'omt', 1, fs.readFileSync(linkUrl).toString());
+    if (otherDocument) {
+        const result = omtDocumentInformationProvider.provideImportsFromDocument(otherDocument);
+        const definedObject = result.localDefinedObject.find(x => x.name == link.name);
```
- [ ] Check if this function can be combined with `getLocationsForLink`

JordyMackay replied:
I refactored this now so that the same function will be used for the current document and for other imported documents. We now have `getLocationsForLink` and `getLocationsInImportedFile` which loads the file and calls `getLocationsForLink`.

JordyMackay commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,365 @@
+    return documentLinks;
+}
+
+export function getImportsFromDocument(document: TextDocument, shorthands?: Map<string, string>) {
```
- [ ] rename function (it returns imports and declared objects)

JordyMackay replied:
Renamed to `getAvailableObjectsFromDocument`

emielb commented on CHANGELOG.md:
```diff
@@ -7,6 +7,12 @@ Check [Keep a Changelog](http://keepachangelog.com/) for recommendations
 on how to structure this file.
 This plugin uses [Semantic versioning](https://semver.org).
 
+## [2.9.0] - 2021-11-23
```
To make this work, we need to make sure the commit message when merging includes either "feat" or "minor". See also https://github.com/phips28/gh-action-bump-version which is used in the CD workflow

emielb replied:
It should be sufficient to have one commit in this PR that includes either of those keywords.

JordyMackay replied:
I added a commit with the name `feature: implemented Go to Definition` :+1:

emielb replied:
Should work :)

emielb commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,359 @@
+        return this.getOmtDocumentInformation(document, shorthands);
+    }
+
+    provideAvailableObjectsFromDocument(document: TextDocument): OmtAvailableObjects {
```
Please add JSDoc comments to your functions

emielb commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,359 @@
+    const documentLinks: OmtLocalObject[] = [];
+
+    declaredObjects.filter(declaredObject => line.includes(declaredObject)).forEach(declaredObject => {
+        const characterIndex = line.search(new RegExp(`${declaredObject}(?=[^a-zA-Z0-9]|$)`)) ?? line.indexOf(declaredObject);
```
You'd probably want to check that the match begins at a word boundary or @ so that `someQuery` does not show up as a match for `bothersomeQuery`.

JordyMackay replied:
I think I can use the same construction I use that the end of the string? So that the character before the declaredObject is not a letter, nor a number?

emielb commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,359 @@
+    const documentLinks: OmtLocalObject[] = [];
+
+    declaredObjects.filter(declaredObject => line.includes(declaredObject)).forEach(declaredObject => {
+        const characterIndex = line.search(new RegExp(`${declaredObject}(?=[^a-zA-Z0-9]|$)`)) ?? line.indexOf(declaredObject);
```
Why is there a `line.search()` and then, if the result is `null` or `undefined`, a call to `line.indexOf()`? `line.search()` returns an index, or `-1`, so the latter is never called. Please remove it

JordyMackay replied:
I wanted to have an alternative for if the RegExp was incomplete, so `line.indexOf()` was used as plan B. I will change this so that that the regex is used for filtering so that plan B is not needed anymore.

emielb commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,359 @@
+    const yamlDocument = parse(document.getText());
+    const availableImports: OmtImport[] = [];
+    const definedObjects: OmtLocalObject[] = [];
+    if ("import" in yamlDocument && shorthands) {
```
It really would be good to combine this reading of the imports with the other one further back in this file. That would save many lines of code. The way that is used here, using the yaml parser, is preferable in my opinion

emielb commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,359 @@
+    const yamlDocument = parse(document.getText());
+    const availableImports: OmtImport[] = [];
+    const definedObjects: OmtLocalObject[] = [];
+    if ("import" in yamlDocument && shorthands) {
```
We really should try to combine this reading of the imports with the other one further back in this file. That would save a lot of code. I think the use of the yaml parser here is better than just parsing the file as text

JordyMackay replied:
I agree, I also prefer the yaml parser, when using that we don't have to use regular expressions anymore with three different groups (whitespace followed by quotes followed by text). It will make it a lot easier to understand.

- [ ] create document links using the YAML parser

emielb commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,359 @@
+
+function findDefinedObjects(value: string, documentText: string, define: string): OmtLocalObject[] {
+    const localDefinedObjects: OmtLocalObject[] = [];
+    const queriesString: any = value;
```
Why do you assign the value parameter to a constant here, and why is it of type any, when the parameter is a string?

JordyMackay replied:
It was not necessary to have this construction, I will remove it :+1:

emielb commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,359 @@
+    return { availableImports, definedObjects };
+}
+
+function findModelEntries(modelEntries: any, documentText: string): OmtLocalObject[] {
```
This any gives rise to compilation warnings, do you thing we can make this into an `object` maybe?

JordyMackay replied:
The result of `parse(document.getText())` is `any` as you can see here: https://eemeli.org/yaml/v1/#parse-amp-stringify. I will check if I can use `Object` instead.

JordyMackay commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,359 @@
+        }
+        catch (error) {
+            if (error instanceof YAMLError) {
+                console.log(error);
```
I think it was because of a test that uses the textFixture `imports.omt`. There are multiple YAML errors in that file, too many spaces online 4, and the url at line 6 should be encapsulated with quotes for example. 

If there are errors like that (and OMT can handle it) I think it is better to skip the YAML validation and ignore the YAMLError.

JordyMackay commented on server/src/omtDocumentInformationProvider.ts:
```diff
@@ -0,0 +1,359 @@
+import { dirname, isAbsolute, resolve } from 'path';
```
I splitted it in three different parts:
- omtDocumentInformationProvider -> getting the document links (imports) and find all used objects in code
- omtAvailableObjectsProvider -> getting only the imports and declared objects
- importMatch -> for parsing the urls in the imports

JordyMackay commented:
@emielb I refactored most of the code and split it up into multiple files. I also removed support for declared imports, otherwise I had to make a temporary solution for something that will be removed in the near future.

emielb commented on README-DEVELOP.md:
```diff
@@ -54,8 +53,9 @@ by the extension in `/client`. The features currently implemented are:
-  - Declared OMT imports
+- Go to Definition
+  - Works for imported and declared objects
+  - Based on text, 
```
```suggestion
  - Based on text
```

emielb commented on server/testFixture/one/imports.omt:
```diff
@@ -1,17 +1,15 @@
+    ../relative.omt:
     -   RelativeAction
-    @one/shorthanded.omt:
+    '@one/shorthanded.omt':
```
The whole point of this line is to check whether the plugin can handle the unquoted version. If that is no longer something the plugin should be able to handle, remove this entry.

JordyMackay replied:
When having imports unquoted, it is not valid YAML anymore. I will remove this entry.

emielb commented on CHANGELOG.md:
```diff
@@ -7,6 +7,16 @@ Check [Keep a Changelog](http://keepachangelog.com/) for recommendations
 on how to structure this file.
 This plugin uses [Semantic versioning](https://semver.org).
 
+## [2.9.0] - 2021-11-29
```
The date should be updated before merging

emielb reviewed (changes requested):
Just a few minor changes
